File: Syntactic.rb
# encoding: utf-8
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# Syntactic.rb

require "./Token.rb"
require "./NonTerminal.rb"
require "./Lexical.rb"
require "./LRTable.rb"
require "./Logger.rb"

class Syntactic
    attr_reader :correctSyntax, :stateStack, :syntaxTree
    attr_writer :stateStack, :syntaxTree
    
    private 
        def getToken()
            @token = @lexAnalyzer.nextToken()
            if @token != nil
                if /^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w\.-]*)*\/?$/.match( @token.symbol )
                    @token.type = Token.getTokenTypes().types["Url"]
                end
            end
        end
    
    public
        def initialize( lex, table )
            @lexAnalyzer = lex
            @token = nil
            @correctSyntax = true
            @stateStack = Array.new()
            @syntaxTree = Array.new()
            @lrTable = table
        end
        
        def analyze()
            done = false
            
            @stateStack.clear()
            @syntaxTree.clear()
            @stateStack.push( 0 )
            
            getToken()
            while @correctSyntax and not done
                if @token != nil
                    Logger.instance.writeLog( "Token: #{ @token.symbol }" )
                    Logger.instance.writeLog( "Type: #{ @token.type }" )
                    done = doAction( @stateStack.last(), @token.type )
                else
                    Logger.instance.writeLog( "Token: $" ) 
                    Logger.instance.writeLog( "Type: End of Input" )
                    done = doAction( @stateStack.last(), Token.getTokenTypes().types["InputEnd"] )
                end
            end
            
            if @correctSyntax
                Logger.instance.writeLog( "Ok" )
            end
        end
        
        def doAction( state, input )
            key = [state, input]
            done = false
            
            if @lrTable.table.has_key?( key )
                value = @lrTable.table[key]
                action = value[0]
                index = value[1]
                done = applyAction( action, index, input )
            else
                Logger.instance.writeLog( "Error: Could not find pair." )
                @correctSyntax = false
            end
            
            return done
        end
        
        def applyAction( action, index, input )
            done = false
            
            if action == LRTable::SHIFT
                Logger.instance.writeLog( "Shift: #{ index }" )
                @stateStack.push( index )
                @syntaxTree.push( @token )
                
                getToken()
            elsif action == LRTable::REDUCE
                Logger.instance.writeLog( "Reduce: #{ index }" )
                applyReduction( index )
            elsif action == LRTable::CHANGE
                Logger.instance.writeLog( "Error: Unreacheable code reached." )
                correctSyntax = false
            elsif action == LRTable::ACCEPT
                done = true
            end
            
            return done
        end
        
        def applyReduction( index )
            times = [3, 4, 5, 3, 2, 2, 0, 3, 3, 3, 2, 2, 2, 2, 0,
                     3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2,
                     0, 2, 0]
            
            @stateStack.pop( times[index - 1] )
            children = @syntaxTree.pop( times[index - 1] )
            
            nonTermKey = case index
                when 1 then "rss_tag"
                when 2 then "rss_start"
                when 3 then "version"
                when 4 then "rss_end"
                when 5 then "rss_children"
                when 6..7 then "channels"
                when 8 then "channel_tag"
                when 9 then "channel_start"
                when 10 then "channel_end"
                when 11..15 then "channel_children"
                when 16 then "title_tag"
                when 17 then "title_start"
                when 18 then "title_end"
                when 19 then "link_tag"
                when 20 then "link_start"
                when 21 then "link_end"
                when 22 then "description_tag"
                when 23 then "description_start"
                when 24 then "description_end"
                when 25 then "item_tag"
                when 26 then "item_start"
                when 27 then "item_end"
                when 28..31 then "item_children"
                when 32..33 then "content"
            end
            nonTerm = NonTerminal.getTypes().types[nonTermKey]
            
            key = [@stateStack.last(), nonTerm]
            if @lrTable.table.has_key?( key )
                @stateStack.push( @lrTable.table[key][1] )
                @syntaxTree.push( NonTerminal.new( nonTerm, children ) )
            else
                Logger.instance.writeLog( "Error: Could not apply reduction." )
                @correctSyntax = false
            end
        end
end

File: Syntactic.py
# encoding: utf-8
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# Syntactic.py

import Token
import NonTerminal
import Lexical
import LRTable
import Logger
import re

class Syntactic( object ):
    
    def __init__( self, lex, table, log ):
        self.lexAnalyzer = lex
        self.token = None
        self.correctSyntax = True
        self.stateStack = list()
        self.syntaxTree = list()
        self.lrTable = table
        self.log = log
        
    def analyze( self ):
        done = False
        
        self.stateStack = list()
        self.syntaxTree = list()
        self.stateStack.append( 0 )
        
        self.getToken()
        while self.correctSyntax and not done:
            if self.token is not None:
                self.log.writeDebug( "Token: " + str( self.token.symbol ) )
                self.log.writeDebug( "Type: " + str( self.token.nodeType ) )
                done = self.doAction( self.stateStack[-1], self.token.nodeType )
            else:
                self.log.writeDebug( "Token: $" )
                self.log.writeDebug( "Type: End of Input" )
                done = self.doAction( self.stateStack[-1], Token.getTokenTypes().types["InputEnd"] )
        
        if self.correctSyntax:
            self.log.writeDebug( "Ok" )
            
    def doAction( self, state, tokenType ):
        key = (state, tokenType)
        done = False
        
        if self.lrTable.table.has_key( key ):
            value = self.lrTable.table[key]
            action = value[0]
            index = value[1]
            done = self.applyAction( action, index, tokenType )
        else:
            self.log.writeError( "Error: Could not find pair." )
            self.log.writeError( str( state ) + ", " + str( tokenType ) )
            self.correctSyntax = False
            
        return done
        
    def applyAction( self, action, index, tokenType ):
        done = False
        
        if action == LRTable.LRTable.SHIFT:
            self.log.writeDebug( "Shift " + str( index ) )
            self.stateStack.append( index )
            self.syntaxTree.append( self.token )
            
            self.getToken()
        elif action == LRTable.LRTable.REDUCE:
            self.log.writeDebug( "Reduce: " + str( index ) )
            self.applyReduction( index )
        elif action == LRTable.LRTable.CHANGE:
            self.log.writeError( "Error: Unreacheable code reacher." )
            self.correctSyntax = False
        elif action == LRTable.LRTable.ACCEPT:
            done = True
            
        return done
        
    def applyReduction( self, index ):
        times = [3, 4, 5, 3, 2, 2, 0, 3, 3, 3, 2, 2, 2, 2, 0,
                 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2,
                 0, 2, 0]
                 
        children = []
        for i in range( 0, times[index - 1] ):
            self.stateStack.pop()
            children.append( self.syntaxTree.pop() )
        children.reverse()

        if index == 1:
            nonTermKey = "rss_tag" 
        elif index == 2:
            nonTermKey = "rss_start"
        elif index == 3:
            nonTermKey = "version"
        elif index == 4:
            nonTermKey = "rss_end"
        elif index == 5:
            nonTermKey = "rss_children"
        elif index == 6 or index == 7:
            nonTermKey = "channels"
        elif index == 8:
            nonTermKey = "channel_tag"
        elif index == 9:
            nonTermKey = "channel_start"
        elif index == 10:
            nonTermKey = "channel_end"
        elif index >= 11 and index <= 15:
            nonTermKey = "channel_children"
        elif index == 16:
            nonTermKey = "title_tag"
        elif index == 17:
            nonTermKey = "title_start"
        elif index == 18:
            nonTermKey = "title_end"
        elif index == 19:
            nonTermKey = "link_tag"
        elif index == 20:
            nonTermKey = "link_start"
        elif index == 21:
            nonTermKey = "link_end"
        elif index == 22:
            nonTermKey = "description_tag"
        elif index == 23:
            nonTermKey = "description_start"
        elif index == 24:
            nonTermKey = "description_end"
        elif index == 25:
            nonTermKey = "item_tag"
        elif index == 26:
            nonTermKey = "item_start"
        elif index == 27:
            nonTermKey = "item_end"
        elif index >= 28 and index <= 31:
            nonTermKey = "item_children"
        elif index == 32 or index == 33:
            nonTermKey = "content"

        nonTerm = NonTerminal.getTypes().types[nonTermKey]

        key = (self.stateStack[-1], nonTerm)
        if self.lrTable.table.has_key( key ):
            self.stateStack.append( self.lrTable.table[key][1] )
            self.syntaxTree.append( NonTerminal.NonTerminal( nonTerm, children ) )
        
    def getToken( self ):
        self.token = self.lexAnalyzer.nextToken()
        if self.token is not None:
            if re.match( "^(https?:\/\/)?([\w\.-]+)\.([\w\.]{2,6})([\/\w\.-]*)\/?$", self.token.symbol ) is not None:
                self.token.nodeType = Token.getTokenTypes().types["Url"]
